File: source code/fontend/ai-powered-st/src/chapter-10/models/requests/request-models.ts
// models/requests/metric-design-request.ts
export interface MetricDesignRequest {
    businessObjectives: string[];
    testingActivities: string[];
    designPrinciples: string[];
    constraints: MetricConstraints;
}

export interface MetricConstraints {
    maxMetrics: number;
    minDataPoints: number;
    requiredDimensions?: string[];
    allowCompositeMetrics: boolean;
}

// models/requests/health-score-request.ts
export interface HealthScoreRequest {
    metricValues: MetricValue[];
    historicalBaselines: HistoricalBaseline[];
    normalizationMethod: string;
    weightingStrategy: string;
    confidenceThreshold: number;
}

export interface MetricValue {
    metricId: string;
    metricName: string;
    value: number;
    timestamp: Date;
    attributes?: { [key: string]: any };
}

export interface HistoricalBaseline {
    metricId: string;
    mean: number;
    standardDeviation: number;
    min: number;
    max: number;
    sampleSize: number;
    periodStart: Date;
    periodEnd: Date;
}

// models/requests/prediction-request.ts
export interface PredictionRequest {
    currentMetrics: MetricValue[];
    historicalTrends: HistoricalTrend[];
    predictionHorizon: number; // Days
    confidenceIntervals: number[];
    includeInterventions: boolean;
}

export interface HistoricalTrend {
    date: Date;
    metricValues: { [key: string]: number };
    seasonality?: string;
}

// models/requests/insight-request.ts
export interface InsightRequest {
    metrics: MetricValue[];
    insightTypes: string[];
    actionabilityThreshold: number;
    context: InsightContext;
}

export interface InsightContext {
    stakeholders?: string[];
    businessGoals?: string[];
    constraints?: { [key: string]: any };
    priorityLevel?: string;
}

// models/requests/optimization-request.ts
export interface OptimizationRequest {
    currentMetrics: MetricDefinition[];
    resourceConstraints: ResourceConstraint[];
    optimizationGoals?: string[];
    preservationRules?: PreservationRule[];
}

export interface MetricDefinition {
    metricId: string;
    name: string;
    category: string;
    collectionMethod: string;
    collectionCost: number;
    businessValue: number;
    dependencies?: string[];
}

export interface ResourceConstraint {
    resourceType: string;
    maxAllocation: number;
    unit: string;
    period: string;
}

export interface PreservationRule {
    metricId: string;
    rule: string;
    reason: string;
}

File: source code/backend/ai-powered-software-testing/Chapter 10/Models/Requests/RequestModels.cs

namespace Chapter_10.Models.Requests
{
    public class MetricDesignRequest
    {
        public string[] BusinessObjectives { get; set; }
        public string[] TestingActivities { get; set; }
        public string[] DesignPrinciples { get; set; }
        public MetricConstraints Constraints { get; set; }
    }

    public class MetricConstraints
    {
        public int MaxMetrics { get; set; }
        public int MinDataPoints { get; set; }
        public string[] RequiredDimensions { get; set; }
        public bool AllowCompositeMetrics { get; set; }
    }
}

// Models/Requests/HealthScoreRequest.cs
namespace Chapter_10.Models.Requests
{
    public class HealthScoreRequest
    {
        public MetricValue[] MetricValues { get; set; }
        public HistoricalBaseline[] HistoricalBaselines { get; set; }
        public string NormalizationMethod { get; set; }
        public string WeightingStrategy { get; set; }
        public double ConfidenceThreshold { get; set; }
    }

    public class MetricValue
    {
        public string MetricId { get; set; }
        public string MetricName { get; set; }
        public double Value { get; set; }
        public DateTime Timestamp { get; set; }
        public Dictionary<string, object> Attributes { get; set; }
    }

    public class HistoricalBaseline
    {
        public string MetricId { get; set; }
        public double Mean { get; set; }
        public double StandardDeviation { get; set; }
        public double Min { get; set; }
        public double Max { get; set; }
        public int SampleSize { get; set; }
        public DateTime PeriodStart { get; set; }
        public DateTime PeriodEnd { get; set; }
    }
}

// Models/Requests/PredictionRequest.cs
namespace Chapter_10.Models.Requests
{
    public class PredictionRequest
    {
        public MetricValue[] CurrentMetrics { get; set; }
        public HistoricalTrend[] HistoricalTrends { get; set; }
        public int PredictionHorizon { get; set; } // Days
        public double[] ConfidenceIntervals { get; set; }
        public bool IncludeInterventions { get; set; }
    }

    public class HistoricalTrend
    {
        public DateTime Date { get; set; }
        public Dictionary<string, double> MetricValues { get; set; }
        public string Seasonality { get; set; }
    }
}

// Models/Requests/InsightRequest.cs
namespace Chapter_10.Models.Requests
{
    public class InsightRequest
    {
        public MetricValue[] Metrics { get; set; }
        public string[] InsightTypes { get; set; }
        public double ActionabilityThreshold { get; set; }
        public InsightContext Context { get; set; }
    }

    public class InsightContext
    {
        public string[] Stakeholders { get; set; }
        public string[] BusinessGoals { get; set; }
        public Dictionary<string, object> Constraints { get; set; }
        public string PriorityLevel { get; set; }
    }
}

// Models/Requests/OptimizationRequest.cs
namespace Chapter_10.Models.Requests
{
    public class OptimizationRequest
    {
        public MetricDefinition[] CurrentMetrics { get; set; }
        public ResourceConstraint[] ResourceConstraints { get; set; }
        public string[] OptimizationGoals { get; set; }
        public PreservationRule[] PreservationRules { get; set; }
    }

    public class MetricDefinition
    {
        public string MetricId { get; set; }
        public string Name { get; set; }
        public string Category { get; set; }
        public string CollectionMethod { get; set; }
        public double CollectionCost { get; set; }
        public double BusinessValue { get; set; }
        public string[] Dependencies { get; set; }
    }

    public class ResourceConstraint
    {
        public string ResourceType { get; set; }
        public double MaxAllocation { get; set; }
        public string Unit { get; set; }
        public string Period { get; set; }
    }

    public class PreservationRule
    {
        public string MetricId { get; set; }
        public string Rule { get; set; }
        public string Reason { get; set; }
    }
}